pkg/standalone-utils/test/VaultActionsRelayer.setup.ts

```diff
@@ -0,0 +1,167 @@
+import { MAX_UINT256, ZERO_ADDRESS } from '@balancer-labs/v2-helpers/src/constants';
+import { deploy, deployedAt } from '@balancer-labs/v2-helpers/src/contract';
+import { actionId } from '@balancer-labs/v2-helpers/src/models/misc/actions';
+import { BigNumberish } from '@balancer-labs/v2-helpers/src/numbers';
+import { Account } from '@balancer-labs/v2-helpers/src/models/types/types';
+import { Dictionary } from 'lodash';
+import TypesConverter from '@balancer-labs/v2-helpers/src/models/types/TypesConverter';
+import TokenList from '@balancer-labs/v2-helpers/src/models/tokens/TokenList';
+import Token from '@balancer-labs/v2-helpers/src/models/tokens/Token';
+import { SwapKind } from '@balancer-labs/balancer-js';
+
+export enum PoolKind {
+  WEIGHTED = 0,
+  LEGACY_STABLE,
+  COMPOSABLE_STABLE,
+  COMPOSABLE_STABLE_V2,
+}
+
+export async function setupRelayerEnvironment(): Promise<{
+  user: SignerWithAddress;
+  other: SignerWithAddress;
+  vault: Vault;
+  relayer: Contract;
+  relayerLibrary: Contract;
+}> {
+  const [, admin, user, other] = await ethers.getSigners();
+
+  // Deploy Balancer Vault
+  const vault = await Vault.create({ admin });
+
+  // Deploy Relayer
+  const relayerLibrary = await deploy('MockBatchRelayerLibrary', { args: [vault.address, ZERO_ADDRESS, ZERO_ADDRESS] });
+  const relayer = await deployedAt('BalancerRelayer', await relayerLibrary.getEntrypoint());
+
+  // Authorize Relayer for all actions
+  const relayerActionIds = await Promise.all(
+    ['swap', 'batchSwap', 'joinPool', 'exitPool', 'setRelayerApproval', 'manageUserBalance'].map((action) =>
+      actionId(vault.instance, action)
+    )
+  );
+  await vault.grantPermissionsGlobally(relayerActionIds, relayer);
+
+  // Approve relayer by sender
+  await vault.setRelayerApproval(user, relayer, true);
+
+  return { user, other, vault, relayer, relayerLibrary };
+}
+
+export async function encodeJoinPool(
+  vault: Vault,
+  relayerLibrary: Contract,
+  params: {
+    poolId: string;
+    userData: string;
+    outputReference?: BigNumberish;
+    sender: Account;
+    recipient: Account;
+    poolKind?: number;
+  }
+): Promise<string> {
+  const { tokens } = await vault.getPoolTokens(params.poolId);
+
+  return relayerLibrary.interface.encodeFunctionData('joinPool', [
+    params.poolId,
+    params.poolKind ?? PoolKind.WEIGHTED,
+    TypesConverter.toAddress(params.sender),
+    TypesConverter.toAddress(params.recipient),
+    {
+      assets: tokens,
+      maxAmountsIn: new Array(tokens.length).fill(MAX_UINT256),
+      userData: params.userData,
+      fromInternalBalance: false,
+    },
+    0,
+    params.outputReference ?? 0,
+  ]);
+}
+
+export async function encodeExitPool(
+  vault: Vault,
+  relayerLibrary: Contract,
+  tokens: TokenList,
+  params: {
+    poolId: string;
+    userData: string;
+    toInternalBalance: boolean;
+    outputReferences?: Dictionary<BigNumberish>;
+    sender: Account;
+    recipient: Account;
+    poolKind?: number;
```

Comment:
I'd not make this argument optional

Reply:
Yes; we did just get burned by that :) Originally it was to make minimal changes (i.e., not change VaultActions.test.ts; just add the new file). But it required a lot of changes anyway, so might as well add one more.